### PR TITLE
iOS README - Route and Directions

### PR DESCRIPTION
**Pull request by vquach2404**
I also changed the title of "Find service area interactive" to "Find service area (interactive)" to include parentheses because "Find closest facility to multiple incidents (service)" has parentheses and I found it clarifying to include them. If anyone opposes, I'll remove them. 

I used syntax `_.` similar to the common sample design where in actual code, included specific variable names. Let me know what you think and if I should change that.

### CODE REVIEW

File: arcgis-ios-sdk-samples/Route & Directions/Find closest facility to multiple incidents (service)/README.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 Find routes from several locations to the respective closest facility.
 
-![Screenshot of Find Closest Facility to Multiple Incidents Service sample](image1.png)
+![Screenshot of Find Closest Facility to Multiple Incidents Service sample](find-closest-facility-service.png)
```
Comment (yo1995) [inline arcgis-ios-sdk-samples/Route & Directions/Find closest facility to multiple incidents (service)/README.md]: Oops. I didn't rename the image files for #838 and #842 . Should we rename the images to reflect the name of the example? @philium

File: arcgis-ios-sdk-samples/Route & Directions/Offline routing/README.md [hunk 2]
```diff
@@ -1,20 +1,43 @@
...
-You tap on the map to add a stop. As you add individual stops the sample solves the route between the stops. If you tap, hold and move then the route will be calculated dynamically; on the fly. You can also switch between the fastest or the shortest route.
+Tap near a road to start adding a stop to the route, tap again to place it on the map. A number graphic will show its order in the route. After adding at least 2 stops, a route will display. Choose "Fastest" or "Shortest" to control how the route is optimized. The route will update on-the-fly while adding stops. 
+
+## How it works
 
+1. Create the map's `AGSBasemap` from a local tile package using an `AGSTileCache` and `AGSArcGISTiledLayer`.
+2. Create an `AGSRouteTask` with an offline locator geodatabase.
+3. Get the `AGSRouteParameters` using `_.getDefaultParameters()`.
+4. Create `AGSStop`s and add them to the route task's parameters.
+5. Solve the route using `routeTask.solveRoute(with:completion:)`.
```
Comment (yo1995) [inline arcgis-ios-sdk-samples/Route & Directions/Offline routing/README.md]:
```suggestion
5. Solve the route using `AGSRouteTask.solveRoute(with:completion:)`.
```
Not sure if you know it, but Philip taught me a very useful shortcut to copy the qualified symbol name for the iOS SDK.

Press <kbd>Control</kbd>+<kbd>Command</kbd>+<kbd>Shift</kbd>+<kbd>Option</kbd>+<kbd>C</kbd> together when a symbol is selected, and it will copy the full qualified symbol name for it! 😏

File: arcgis-ios-sdk-samples/Route & Directions/Find route/README.md [hunk 3]
```diff
@@ -1,20 +1,36 @@
...
+For simplicity, the sample comes loaded with a start and end stop. You can click on the Find Route button to display a route between these stops. Once the route is generated, tap the directions icon for turn-by-turn directions shown in a list.
 
 ## How it works
 
-The sample uses the `defaultRouteParameters(completion:)` method on `AGSRouteTask` to get the default parameters from the service. Sets the stops in those parameters and the `returnDirections` to true. Then uses the `solveRoute(with:completion:)` method to solve for the route. Once the route is in, the sample displays the list of directions using the `directionManeuvers` property on the `AGSRoute` returned in the `AGSRouteResult`
+1. Create a `AGSRouteTask` using a URL to an online route service.
+2. Generate default `AGSRouteParameters` using `_.getDefaultParameters`.
+3. Set `_.returnDirections` on the parameters to true.
+4. Add `AGSStop`s to the parameters `stops` collection for each destination.
+5. Solve the route using `routeTask.solveRoute(with:completion:)` to get an `AGSRouteResult`.
```
Comment (yo1995) [inline arcgis-ios-sdk-samples/Route & Directions/Find route/README.md]:
```suggestion
5. Solve the route using `AGSRouteTask.solveRoute(with:completion:)` to get an `AGSRouteResult`.
```

File: arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md [hunk 4]
```diff
@@ -1,20 +1,73 @@
...
+    iv. Add the graphic to the barriers overlay.
+5. When ready to find the route, configure the route parameters.
+    i. Set the `ReturnStops` and `ReturnDirections` to `true`.
+    ii. Create an `AGSStop` for each graphic in the stops graphics overlay. Add that stop to a list, then call `_routeParameters.setStops(stops:)`.
+    iii. Create a `AGSPolygonBarrier` for each graphic in the barriers graphics overlay. Add that barrier to a list, then call `_routeParameters.setPolygonBarriers(polygonBarriers:)`.
+    iv. If the user will accept routes with the stops in any order, set `FindBestSequence` to `true` to find the most optimal route.
+    v. If the user has a definite start point, set `AGSRouteParameters.preserveFirstStop` to `true`.
+    vi. If the user has a definite final destination, set `AGSRouteParameters.preserveLastStop` to `true`.
+6. Calculate and display the route.
+    i. Call `_routeTask.solveRoute(with:completion:)` to get an `AGSRouteResult`.
```
Comment (yo1995) [inline arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md]:
```suggestion
    i. Call `AGSRouteTask.solveRoute(with:completion:)` to get an `AGSRouteResult`.
```

File: arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md [hunk 5]
```diff
@@ -1,20 +1,73 @@
...
+2. Get the default route parameters for the service by calling `AGSRouteTask.getDefaultParameters`.
+3. When the user adds a stop, add it to the route parameters.
+    i. Normalize the geometry; otherwise the route job would fail if the user included any stops over the 180th degree meridian.
+    ii. Get the name of the stop by counting the existing stops - `.stopGraphicsOverlay.graphics.index(of:) + 1`.
+    iii. Create a composite symbol for the stop. This sample uses a blue marker and a text symbol.
+    iv. Create the graphic from the geometry and the symbol.
+    v. Add the graphic to the stops graphics overlay.
+4. When the user adds a barrier, create a polygon barrier and add it to the route parameters.
+    i. Normalize the geometry (see **3i** above).
+    ii. Buffer the geometry to create a larger barrier from the tapped point by calling `AGSGeometryEngine.bufferGeometry(geometry:byDistance:)`.
```
Comment (yo1995) [inline arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md]:
```suggestion
    ii. Buffer the geometry to create a larger barrier from the tapped point by calling `class AGSGeometryEngine.bufferGeometry(geometry:byDistance:)`.
```
Static methods should include the `class` part, and it will be copied if you use the copy qualified symbol name shortcut mentioned above.

File: arcgis-ios-sdk-samples/Route & Directions/Offline routing/README.md (hunk 2 above)
Comment (vquach2404) [reply arcgis-ios-sdk-samples/Route & Directions/Offline routing/README.md]: Oh, I didn't know that. Thanks! Very helpful 🙂

File: arcgis-ios-sdk-samples/Route & Directions/Find closest facility to multiple incidents (service)/README.md (hunk 1 above)
Comment (yo1995) [reply arcgis-ios-sdk-samples/Route & Directions/Find closest facility to multiple incidents (service)/README.md:5]: The only worry is for the metadata json file.

In the last sprint review meeting, Mike mentioned that the metadata files will be automatically updated based on the README file with a script. Hopefully these renames won't break anything on the website! I'll rename the images based on example names.

File: arcgis-ios-sdk-samples/Route & Directions/Find route/README.md [hunk 6]
```diff
@@ -16,7 +16,7 @@ For simplicity, the sample comes loaded with a start and end stop. You can click
 ## How it works
 
 1. Create a `AGSRouteTask` using a URL to an online route service.
-2. Generate default `AGSRouteParameters` using `_.getDefaultParameters`.
+2. Generate default `AGSRouteParameters` using `FindRouteViewController.getDefaultParameters()`.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Find route/README.md:19]: This should be the API used to get the default route parameters:

```suggestion
2. Generate default `AGSRouteParameters` using `AGSRouteTask.defaultRouteParameters(completion:)`.
```

File: arcgis-ios-sdk-samples/Route & Directions/Find closest facility to multiple incidents (service)/README.md [hunk 7]
```diff
@@ -24,7 +24,7 @@ Tap the Solve Routes button to solve and display the route from each incident (f
 5. Add the array of all incidents to the task parameters.
 6. Get an `AGSClosestFacilityResult` by solving the task with the provided parameters.
 7. Find the closest facility for each incident by iterating over the array of incidents.
-8. Display the route as an `AGSGraphic`.
+8. Display the route as an `AGSGraphic(geometry:symbol:)`.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Find closest facility to multiple incidents (service)/README.md:27]: Actually, I think how it was before is better.
```suggestion
8. Display the route as an `AGSGraphic`.
```

File: arcgis-ios-sdk-samples/Route & Directions/Find route/README.md [hunk 8]
```diff
@@ -1,20 +1,36 @@
...
 
+![Solved route](find-route-1.png)
+![List of directions](find-route-2.png)
 
-## How to use the sample
-
-For simplicity, the sample comes loaded with a start and end stop. You can tap on the route button to get a route between these stops. Once the route is generated, the directions button gets enabled. Tap on the button to see the step by step directions for the route.
+## Use case
 
+Find routes with driving directions between any number of locations. You might use the ArcGIS platform to create a custom network for routing on a private roads.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Find route/README.md:10]:
```suggestion
Find routes with driving directions between any number of locations. You might use the ArcGIS platform to create a custom network for routing on private roads.
```

File: arcgis-ios-sdk-samples/Route & Directions/Find route/README.md [hunk 9]
```diff
@@ -1,20 +1,36 @@
...
+## How to use the sample
 
+For simplicity, the sample comes loaded with a start and end stop. You can click on the Find Route button to display a route between these stops. Once the route is generated, tap the directions icon for turn-by-turn directions shown in a list.
 
 ## How it works
 
-The sample uses the `defaultRouteParameters(completion:)` method on `AGSRouteTask` to get the default parameters from the service. Sets the stops in those parameters and the `returnDirections` to true. Then uses the `solveRoute(with:completion:)` method to solve for the route. Once the route is in, the sample displays the list of directions using the `directionManeuvers` property on the `AGSRoute` returned in the `AGSRouteResult`
+1. Create a `AGSRouteTask` using a URL to an online route service.
+2. Generate default `AGSRouteParameters` using `FindRouteViewController.getDefaultParameters()`.
+3. Set `_.returnDirections` on the parameters to true.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Find route/README.md:20]:
```suggestion
3. Set `returnDirections` on the parameters to `true`.
```

File: arcgis-ios-sdk-samples/Route & Directions/Find route/README.md [hunk 10]
```diff
@@ -1,20 +1,36 @@
...
+4. Add `AGSStop`s to the parameters `stops` collection for each destination.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Find route/README.md:21]:
```suggestion
4. Create an `ASGStop` object for each destination and assign the stops to the parameters using `AGSRouteParameters.setStops(_:)`.
```

File: arcgis-ios-sdk-samples/Route & Directions/Find route/README.md [hunk 11]
```diff
@@ -1,20 +1,36 @@
...
+5. Solve the route using `AGSRouteTask.solveRoute(with:completion:)` to get an `AGSRouteResult`.
+6. Iterate through the result's `AGSRoute`s. To display the route, create a graphic using the geometry from `_routeGeometry`. To display directions, use `route.getDirectionManeuvers()`, and for each `DirectionManeuver`, display `_.directionManeuvers`.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Find route/README.md:23]:
```suggestion
6. Iterate through the result's `AGSRoute`s. To display the route, create a graphic using the geometry from the route's `routeGeometry`. To display directions, get the direction maneuvers from the route's `directionManeuvers` property, and, for each maneuver, display the maneuver's `directionText`.
```

File: arcgis-ios-sdk-samples.xcodeproj/project.pbxproj [hunk 12]
```diff
@@ -2266,7 +2266,7 @@
 				3E6AE3F31C6D1C920026E59C /* Find route */,
 				3EC39C191C9A5770003F6459 /* Route around barriers */,
 				3E6AE3EA1C6D10F70026E59C /* Offline routing */,
-				3ED75D001ED39EDF009B1F75 /* Find service area interactive */,
+				3ED75D001ED39EDF009B1F75 /* Find service area (interactive) */,
```
Comment (philium) [inline arcgis-ios-sdk-samples.xcodeproj/project.pbxproj:2269]: Renaming a sample is more than just updating a README. Something like this that affects the project file itself should really go in through a PR to `v.next`. Changing the name in the README and the plist is fine, but please revert this group name change and instead submit a separate PR for it to `v.next`.

File: arcgis-ios-sdk-samples/Route & Directions/Offline routing/README.md [hunk 13]
```diff
@@ -1,20 +1,43 @@
...
-You tap on the map to add a stop. As you add individual stops the sample solves the route between the stops. If you tap, hold and move then the route will be calculated dynamically; on the fly. You can also switch between the fastest or the shortest route.
+Tap near a road to start adding a stop to the route, tap again to place it on the map. A number graphic will show its order in the route. After adding at least 2 stops, a route will display. Choose "Fastest" or "Shortest" to control how the route is optimized. The route will update on-the-fly while adding stops. 
+
+## How it works
 
+1. Create the map's `AGSBasemap` from a local tile package using an `AGSTileCache` and `AGSArcGISTiledLayer`.
+2. Create an `AGSRouteTask` with an offline locator geodatabase.
+3. Get the `AGSRouteParameters` using `OfflineRoutingViewController.getDefaultParameters()`.
+4. Create `AGSStop`s and add them to the route task's parameters.
+5. Solve the route using `AGSRouteTask.solveRoute(with:completion:)`.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Offline routing/README.md:22]: See comments on Find Route's README.

File: arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md [hunk 14]
```diff
@@ -1,20 +1,73 @@
...
+## How to use the sample
 
-![](image1.png)
-![](image2.png)
+Tap 'Add stop' to add stops to the route. Tap 'Add barrier' to add areas that can't be crossed by the route. Tap 'Route' to find the route and display it. Tap the settings button to toggle preferences like find the best sequence or preserve the first or last stop. Additionally, tap the directions button to view a list of the directions. 
 
 ## How it works
 
-The sample uses the `defaultRouteParameters(completion:)` method on `AGSRouteTask` to get the default parameters from the service. Barriers are created using the initializer `init(polygon:)` on `AGSPolygonBarrier`. Sets the stops and barriers in those parameters and the `returnDirections` to true. Then uses the `solveRoute(with:completion:)` method to solve for the route. Once the route is in, the sample displays the list of directions using the `directionManeuvers` property on the `AGSRoute` returned in the `AGSRouteResult`.
+1. Create the route task by calling `AGSRouteTask(url:)` with the URL to a Network Analysis route service.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md:20]:
```suggestion
1. Create the route task using `AGSRouteTask.init(url:)` with the URL to a Network Analysis route service.
```

File: arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md [hunk 15]
```diff
@@ -1,20 +1,73 @@
...
+2. Get the default route parameters for the service by calling `AGSRouteTask.getDefaultParameters`.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md:21]: `getDefaultParameters` is not the name of the method on `AGSRouteTask`.

File: arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md [hunk 16]
```diff
@@ -1,20 +1,73 @@
...
+3. When the user adds a stop, add it to the route parameters.
+    i. Normalize the geometry; otherwise the route job would fail if the user included any stops over the 180th degree meridian.
+    ii. Get the name of the stop by counting the existing stops - `.stopGraphicsOverlay.graphics.index(of:) + 1`.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md:24]:
```suggestion
    ii. Get the name of the stop by counting the existing stops - `stopGraphicsOverlay.graphics.index(of:) + 1`.
```

File: arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md [hunk 17]
```diff
@@ -1,20 +1,73 @@
...
+    iii. Create a composite symbol for the stop. This sample uses a blue marker and a text symbol.
+    iv. Create the graphic from the geometry and the symbol.
+    v. Add the graphic to the stops graphics overlay.
+4. When the user adds a barrier, create a polygon barrier and add it to the route parameters.
+    i. Normalize the geometry (see **3i** above).
+    ii. Buffer the geometry to create a larger barrier from the tapped point by calling `class AGSGeometryEngine.bufferGeometry(geometry:byDistance:)`.
+    iii. Create the graphic from the geometry and the symbol.
+    iv. Add the graphic to the barriers overlay.
+5. When ready to find the route, configure the route parameters.
+    i. Set the `ReturnStops` and `ReturnDirections` to `true`.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md:34]:
```suggestion
    i. Set the `returnStops` and `returnDirections` to `true`.
```

File: arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md [hunk 18]
```diff
@@ -1,20 +1,73 @@
...
+    ii. Create an `AGSStop` for each graphic in the stops graphics overlay. Add that stop to a list, then call `_routeParameters.setStops(stops:)`.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md:35]:
```suggestion
    ii. Create an `AGSStop` for each graphic in the stops graphics overlay. Add that stop to a list, then call `AGSRouteParameters.setStops(_:)`.
```

File: arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md [hunk 19]
```diff
@@ -1,20 +1,73 @@
...
+    iii. Create a `AGSPolygonBarrier` for each graphic in the barriers graphics overlay. Add that barrier to a list, then call `_routeParameters.setPolygonBarriers(polygonBarriers:)`.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md:36]:
```suggestion
    iii. Create a `AGSPolygonBarrier` for each graphic in the barriers graphics overlay. Add that barrier to a list, then call `AGSRouteParameters.setPolygonBarriers(_:)`.
```

File: arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md [hunk 20]
```diff
@@ -1,20 +1,73 @@
...
+    iv. If the user will accept routes with the stops in any order, set `FindBestSequence` to `true` to find the most optimal route.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md:37]:
```suggestion
    iv. If the user will accept routes with the stops in any order, set `findBestSequence` to `true` to find the most optimal route.
```

File: arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md [hunk 21]
```diff
@@ -1,20 +1,73 @@
...
+    v. If the user has a definite start point, set `AGSRouteParameters.preserveFirstStop` to `true`.
+    vi. If the user has a definite final destination, set `AGSRouteParameters.preserveLastStop` to `true`.
+6. Calculate and display the route.
+    i. Call `AGSRouteTask.solveRoute(with:completion:)` to get an `AGSRouteResult`.
+    ii. Get the first returned route by calling `AGSRouteResult.routes.first()`.
+    iii. Get the geometry from the route as a polyline by accessing the `_.routeGeometry` property.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md:43]:
```suggestion
    iii. Get the geometry from the route as a polyline by accessing the `routeGeometry` property.
```

File: arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md [hunk 22]
```diff
@@ -1,20 +1,73 @@
...
+    iv. Create a graphic from the polyline and a simple line symbol.
+    v. Display the steps on the route, available from `_.directionManeuvers`.
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Route around barriers/README.md:45]:
```suggestion
    v. Display the steps on the route, available from `directionManeuvers`.
```

File: arcgis-ios-sdk-samples/Route & Directions/Find service area interactive/FindServiceAreaInteractive.storyboard [hunk 23]
```diff
@@ -1,9 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="15701" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" useSafeAreas="YES" colorMatched="YES" initialViewController="uF5-61-Oq4">
+<document type="com.apple.InterfaceBuilder3.CocoaTouch.Storyboard.XIB" version="3.0" toolsVersion="15705" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" useSafeAreas="YES" colorMatched="YES" initialViewController="uF5-61-Oq4">
```
Comment (philium) [inline arcgis-ios-sdk-samples/Route & Directions/Find service area interactive/FindServiceAreaInteractive.storyboard:2]: Please revet the changes in this file. They are not related to updating the READMEs.

Comment (vquach2404) [reply arcgis-ios-sdk-samples/Route & Directions/Find service area interactive/FindServiceAreaInteractive.storyboard:2]: I'm not sure how this change happened..  This happens quite often -- somehow changes like these are made in the storyboard without my knowledge. I usually review the changes before I commit them but they are easily overlooked at times. Do you know why or how I can prevent this from happening?

Comment (yo1995) [reply arcgis-ios-sdk-samples/Route & Directions/Find service area interactive/FindServiceAreaInteractive.storyboard:2]: My own suggestion is to avoid touching Storyboards during other development, and do separate commits for them. In this way even if you commit-ed some unwanted changes, it is easier to revert to a previous clean state. 😅 

And often check the diff either locally or on GitHub `Files Changed` tab.

Comment (vquach2404) [reply arcgis-ios-sdk-samples/Route & Directions/Find service area interactive/FindServiceAreaInteractive.storyboard]: @yo1995 I typically don't open the Storyboards during my readme-update process. Sometimes I may select the wrong file on accident but don't purposefully edit anything in that file. So I'm not sure why these Storyboard changes occur. I'm wondering if upon opening the Storyboard, my mouse detects movement and somehow makes changes.

Comment (yo1995) [reply arcgis-ios-sdk-samples/Route & Directions/Find service area interactive/FindServiceAreaInteractive.storyboard]: My guess is when you load any IB files, storyboards, etc with a later version of Xcode, even if you don't change anything there, Xcode would still change the file version to the latest.   
And any tiny interaction with storyboard will definitely change it.  
Better to discard those changes once you see them 😓
